src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -0,0 +1,468 @@
+        return np.asarray(self.predictor.predict_proba(X))
+
+    def get_leaf_matrix(self, scale: bool = False) -> np.ndarray:
+        if not self.root:
+            raise ValueError("The tree has not been trained before trying to predict")
+
+        leaf_nodes = self.leaf_nodes
+        n_obs = self.n_obs
+
+        matrix = np.zeros((n_obs, n_obs))
+        if (not leaf_nodes):  # make sure that there are calculated observations
+            return matrix
+        for node in leaf_nodes:
+            if scale:
+                n_node = node.indices.shape[0]
+                matrix[np.ix_(node.indices, node.indices)] = 1/n_node
+            else:
+                matrix[np.ix_(node.indices, node.indices)] = 1
+
+        return matrix
+
+    def predict_leaf_matrix(self, X: np.ndarray, scale: bool = False):
+        if not self.root:
+            raise ValueError("The tree has not been trained before trying to predict")
+        return self.predictor.predict_leaf_matrix(X, scale)
+
+
+# From below here, it is the DepthTreeBuilder
+class queue_obj:
+    """
+    Queue object for the splitter depthtree builder class
+    """
+
+    def __init__(
+        self,
+        indices: np.ndarray,
+        depth: int,
+        impurity: float,
+        parent: Node | None = None,
+        is_left: bool | None = None,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        indices : np.ndarray
+            indicies waiting to be split
+        depth : int
+            depth of the node which is to be made
+        impurity : float
+            impurity in the node to be made
+        parent : Node | None, optional
+            parent of the node to be made, by default None
+        is_left : bool | None, optional
+            whether the object is left or right, by default None
+        """
+
+        self.indices = indices
+        self.depth = depth
+        self.impurity = impurity
+        self.parent = parent
+        self.is_left = is_left
+
+
+class DepthTreeBuilder:
+    """
+    Depth first tree builder
+    """
+
+    def __init__(
+        self,
+        X: np.ndarray,
+        Y: np.ndarray,
+        int_max_features: int | None,
+        sample_weight: np.ndarray,
+        criteria: Criteria,
+        splitter: Splitter,
+        leaf_builder_class: LeafBuilder,
+        predict_class: Predict,
+        sample_indices: np.ndarray | None = None,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        X : np.ndarray
+            The feature values
+        Y : np.ndarray
+            The response values
+        feature_indices : np.ndarray
+            Which features to use
+        sample_indices : np.ndarray
+            Indicies of the samples to use.
+        criteria : Criteria
+            Criteria class used for impurity calculations
+        splitter : Splitter | None, optional
+            Splitter class used to split data, by default None
+        """
+        self.features = X
```

Comment:
Just noticed this now but should be consistent and always call these variables "X" and "Y" instead of features and response? (as done in predict class for example).

Reply:
Indeed. Should we make it such that everything is X and Y or everything is features and response.

I think this stems from early days, where we wanted it to be specifically those names.

Reply:
I would suggest to go with with the variables "X" and "Y" (or if we want to be closer to sklearn "X" and "y") and only in the documentation refer to them as features and response.